bgpd/bgp_attr.c

```diff
@@ -2044,7 +2029,35 @@ bgp_attr_parse (struct peer *peer, struct attr *attr, bgp_size_t size,
         return ret;
       }
   }
-  
+
+  /*
+   * According to section 6.3 of RFC4271, syntactically incorrect NEXT_HOP
+   * attribute must result in a NOTIFICATION message (this is implemented below).
+   * At the same time, semantically incorrect NEXT_HOP is more likely to be just
+   * logged locally (this is implemented somewhere else). The UPDATE message
+   * gets ignored in any of these cases.
+   *
+   * According to section 1.3 of RFC2858, an UPDATE message that carries no NLRI,
+   * other than the one encoded in the MP_REACH_NLRI attribute, should not carry
+   * the NEXT_HOP attribute. If such a message contains the NEXT_HOP attribute,
+   * the BGP speaker that receives the message should ignore this attribute.
+   */
+  if (CHECK_FLAG (attr->flag, ATTR_FLAG_BIT (BGP_ATTR_NEXT_HOP))
```

Comment:
Why don't put this check inside of bgp_attr_nexthtop() ?

Reply:
in bgp_attr_nexthop(), we may not parse the MP_REACH_NLRI attribute yet. We have to do the check after we have parsed all attributes in the update message.